File: iosApp/iosApp/viewmodel/EntitiesViewModel.swift
import Foundation
import SharedLogic

/// Client-side role filter for the entities list. Lives in the ViewModel layer
/// (not shared) per the ticket.
enum EntitiesFilter: CaseIterable, Identifiable {
    case all, customer, supplier, middleman
    var id: Self { self }
    var label: String {
        switch self {
        case .all: return "All"
        case .customer: return "Customers"
        case .supplier: return "Suppliers"
        case .middleman: return "Middlemen"
        }
    }
    var role: EntityRole? {
        switch self {
        case .all: return nil
        case .customer: return EntityRole.customer
        case .supplier: return EntityRole.supplier
        case .middleman: return EntityRole.middleman
        }
    }
}

/// One merged list row: an entity + its live HL balance (nil until balances load or
/// if HL has no customer yet — e.g. still PENDING).
struct EntityListItem: Identifiable {
    let entity: Entity
    let balance: EntityBalance?
    var id: String { entity.id }
}

@MainActor
final class EntitiesViewModel: ObservableObject {
    // Merged + filtered rows the UI renders.
    @Published private(set) var items: [EntityListItem] = []
    @Published var searchText: String = "" { didSet { recompute() } }
    @Published var filter: EntitiesFilter = .all { didSet { recompute() } }

    @Published private(set) var isLoadingList: Bool = true
    @Published private(set) var isLoadingBalances: Bool = false
    @Published private(set) var balancesError: HlError? = nil
    @Published private(set) var listError: String? = nil
    @Published var actionError: String? = nil

    /// Carry-over #2: a `profiles == NONE` user gets a clean denied state — never a crash.
    @Published private(set) var permissionDenied: Bool = false
    /// True only for `profiles == MANAGE` — gates add/edit/archive in the UI.
    @Published private(set) var canManage: Bool = false

    private let authRepo = FirebaseAuthRepositoryImpl()

    private var session: UserSession? = nil
    private var config: FirebaseClientConfig? = nil
    private var repo: BackendEntityRepository? = nil
    private var ledger: KtorEntityLedgerRepository? = nil   // owns its HttpClient — closed in deinit (carry-over #1)
    private var balancesUseCase: GetEntityBalancesUseCase? = nil
    private var observeTask: Task<Void, Never>? = nil

    private var allEntities: [Entity] = []
    private var balances: [String: EntityBalance] = [:]

    // Bind the active session + config and wire the repo stack via manual DI.
    // Idempotent — only the first bind for a (uid, config) pair wires things up.
    func bind(session: UserSession, config: FirebaseClientConfig) {
        if self.session?.uid == session.uid, self.config == config { return }
        self.session = session
        self.config = config

        // `canManage` is a UI capability (gates add/edit/archive) — not the list rule.
        // The list's permission gate + isActive filter come from ObserveEntitiesUseCase
        // (shared), which throws PermissionDeniedException for profiles == NONE; we catch
        // that below to show the denied state. (Ticket #34: single source of truth.)
        canManage = session.permissions.profiles == PermissionLevel.manage

        let tokens = HlTokenRepositoryImpl(authRepo: authRepo, activeConfig: config)
        let ledger = KtorEntityLedgerRepository(tokenProvider: tokens, baseUrl: AromexConfig.hlBaseURL.absoluteString)
        self.ledger = ledger
        self.balancesUseCase = GetEntityBalancesUseCase(ledger: ledger)
        self.repo = BackendEntityRepository(config: config, uid: session.uid)

        startObserving(session: session)
        loadBalances()
    }

    /// Consume the shared `ObserveEntitiesUseCase` Flow (SKIE `AsyncSequence`). The gate
    /// (profiles == NONE → PermissionDeniedException) and the `isActive` filter both come
    /// from shared code — no duplicated list logic here.
    private func startObserving(session: UserSession) {
        guard let repo = repo else { return }
        observeTask?.cancel()
        isLoadingList = true
        let useCase = ObserveEntitiesUseCase(repository: repo)
        observeTask = Task { [weak self] in
            do {
                let flow = try useCase.execute(session: session, includeArchived: false)
                for try await entities in flow {
                    guard let self = self else { return }
                    self.allEntities = entities.sorted { $0.name.lowercased() < $1.name.lowercased() }
                    self.isLoadingList = false
                    self.listError = nil
                    self.permissionDenied = false
                    self.recompute()
                }
            } catch {
                guard let self = self else { return }
                self.isLoadingList = false
                if self.isPermissionDenied(error) {
                    self.permissionDenied = true
                } else {
                    self.listError = self.describe(error)
                }
            }
        }
    }

    /// Fetch all HL balances once and merge by externalId (== entity id). Called on
    /// bind, on pull-to-refresh, and after a save.
    func loadBalances() {
        guard let useCase = balancesUseCase, let session = session else { return }
        if isLoadingBalances { return }
        isLoadingBalances = true
        balancesError = nil
        Task {
            do {
                let result = try await useCase.execute(session: session)
                self.balances = result
                self.isLoadingBalances = false
                self.balancesError = nil
                self.recompute()
            } catch {
                self.isLoadingBalances = false
                self.balancesError = mapHlError(error)
            }
        }
    }

    /// A single party's fresh balance for the detail screen.
    func fetchBalance(id: String) async -> EntityBalance? {
        guard let useCase = balancesUseCase, let session = session else { return nil }
        return try? await useCase.executeSingle(session: session, externalId: id)
    }

    /// Soft-archive via the shared use case (guards Walk-in + MANAGE). The live
    /// listener drops the row automatically once Firestore reflects isActive=false.
    func archive(_ entity: Entity) {
        guard let session = session, let repo = repo else { return }
        let useCase = ArchiveEntityUseCase(repository: repo)
        Task {
            do {
                try await useCase.execute(session: session, entity: entity)
            } catch {
                self.actionError = describe(error)
            }
        }
    }

    // MARK: - Client-side merge / search / filter (no re-fetch)

    private func recompute() {
        let query = searchText.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let roleFilter = filter.role
        let filtered = allEntities.filter { entity in
            matchesRole(entity, roleFilter) && matchesSearch(entity, query)
        }
        items = filtered.map { EntityListItem(entity: $0, balance: balances[$0.id]) }
    }

    private func matchesRole(_ entity: Entity, _ role: EntityRole?) -> Bool {
        guard let role = role else { return true }
        return entity.roles.contains(role)
    }

    private func matchesSearch(_ entity: Entity, _ query: String) -> Bool {
        if query.isEmpty { return true }
        if entity.name.lowercased().contains(query) { return true }
        if entity.phones.contains(where: { $0.lowercased().contains(query) }) { return true }
        if entity.roles.contains(where: { $0.name.lowercased().contains(query) }) { return true }
        return false
    }

    // MARK: - Error mapping

    private func mapHlError(_ error: Error) -> HlError {
        let ns = error as NSError
        if let ex = ns.userInfo["KotlinException"] as? HlException {
            return ex.error
        }
        return HlError.Unexpected(message: error.localizedDescription)
    }

    /// The observe Flow surfaces a shared `PermissionDeniedException` (profiles == NONE)
    /// as an NSError carrying the Kotlin exception. Detect it to show the denied state.
    private func isPermissionDenied(_ error: Error) -> Bool {
        return (error as NSError).userInfo["KotlinException"] is PermissionDeniedException
    }

    private func describe(_ error: Error) -> String {
        let ns = error as NSError
        if let throwable = ns.userInfo["KotlinException"] as? KotlinThrowable {
            return throwable.message ?? "Something went wrong"
        }
        return ns.localizedDescription
    }

    deinit {
        // Carry-over #1: this VM owns the KtorEntityLedgerRepository's HttpClient, so
        // it MUST close it on clear to avoid a connection leak. Cancelling the observe
        // Task tears down the shared Flow, which detaches the Firestore listener.
        observeTask?.cancel()
        ledger?.close()
    }
}

File: androidApp/src/main/kotlin/com/humblesolutions/aromex/ui/entities/EntitiesViewModel.kt
package com.humblesolutions.aromex.ui.entities

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.humblesolutions.aromex.data.BackendEntityRepository
import com.humblesolutions.aromex.data.FirebaseAuthRepository
import com.humblesolutions.aromex.data.HL_API_BASE_URL
import com.humblesolutions.aromex.data.HlTokenRepository
import com.humblesolutions.aromex.data.KtorEntityLedgerRepository
import com.humblesolutions.aromex.model.Entity
import com.humblesolutions.aromex.model.EntityBalance
import com.humblesolutions.aromex.model.EntityInput
import com.humblesolutions.aromex.model.EntityRole
import com.humblesolutions.aromex.model.FirebaseClientConfig
import com.humblesolutions.aromex.model.PermissionLevel
import com.humblesolutions.aromex.model.UserSession
import com.humblesolutions.aromex.repository.EntityLedgerRepository
import com.humblesolutions.aromex.usecase.ArchiveEntityUseCase
import com.humblesolutions.aromex.usecase.GetEntityBalancesUseCase
import com.humblesolutions.aromex.usecase.ObserveEntitiesUseCase
import com.humblesolutions.aromex.usecase.SaveEntityUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

enum class EntitiesFilter { ALL, CUSTOMER, SUPPLIER, MIDDLEMAN }

/** One list row: the party + its live HL balance (null until balances load / while PENDING). */
data class EntityRow(val entity: Entity, val balance: EntityBalance?)

data class EntitiesUiState(
    val session: UserSession? = null,
    val canManage: Boolean = false,
    val noAccess: Boolean = false,
    val rows: List<EntityRow> = emptyList(),
    val query: String = "",
    val filter: EntitiesFilter = EntitiesFilter.ALL,
    val isLoadingBalances: Boolean = false,
    val error: String? = null,
    val saving: Boolean = false,
    val saveError: String? = null,
)

/**
 * List + add/edit/archive for Profiles/Entities (test-only UI).
 *
 * The ViewModel is the cache: it holds the live entity list (Firestore snapshot
 * listener via [ObserveEntitiesUseCase]) and the HL balances, merges them by
 * externalId (== entity id), and does search/filter client-side. Balances are
 * re-fetched on demand (HL has no push).
 */
class EntitiesViewModel(application: Application) : AndroidViewModel(application) {

    private val authRepo = FirebaseAuthRepository(application)

    private var entityRepo: BackendEntityRepository? = null
    private var ledgerRepo: EntityLedgerRepository? = null
    private var observeUseCase: ObserveEntitiesUseCase? = null
    private var balancesUseCase: GetEntityBalancesUseCase? = null
    private var archiveUseCase: ArchiveEntityUseCase? = null
    private var saveUseCase: SaveEntityUseCase? = null

    private var session: UserSession? = null
    private var config: FirebaseClientConfig? = null

    private var entities: List<Entity> = emptyList()
    private var balances: Map<String, EntityBalance> = emptyMap()

    private val _uiState = MutableStateFlow(EntitiesUiState())
    val uiState: StateFlow<EntitiesUiState> = _uiState.asStateFlow()

    fun bind(session: UserSession, config: FirebaseClientConfig) {
        if (this.session?.uid == session.uid && this.config == config) return
        this.session = session
        this.config = config

        val entRepo = BackendEntityRepository(getApplication(), config)
        val tokens = HlTokenRepository(authRepo = authRepo, activeConfig = config)
        val ledger = KtorEntityLedgerRepository(tokens, HL_API_BASE_URL)
        entityRepo = entRepo
        ledgerRepo = ledger
        observeUseCase = ObserveEntitiesUseCase(entRepo)
        balancesUseCase = GetEntityBalancesUseCase(ledger)
        archiveUseCase = ArchiveEntityUseCase(entRepo)
        saveUseCase = SaveEntityUseCase(entRepo)

        val profiles = session.permissions.profiles
        if (profiles == PermissionLevel.NONE) {
            _uiState.update { it.copy(session = session, noAccess = true) }
            return
        }
        _uiState.update {
            it.copy(session = session, canManage = profiles == PermissionLevel.MANAGE, noAccess = false)
        }
        observeEntities()
        refreshBalances()
    }

    private fun observeEntities() {
        val useCase = observeUseCase ?: return
        val current = session ?: return
        viewModelScope.launch {
            // ObserveEntitiesUseCase throws PermissionDeniedException at call time for NONE — we
            // already gate on that in bind(), and collect here so a stream error can't crash init.
            runCatching {
                useCase.execute(current)
                    .catch { e -> _uiState.update { it.copy(error = e.message ?: "Failed to load parties") } }
                    .collect { list ->
                        entities = list
                        recompute()
                    }
            }.onFailure { e -> _uiState.update { it.copy(error = e.message) } }
        }
    }

    fun refreshBalances() {
        val useCase = balancesUseCase ?: return
        val current = session ?: return
        _uiState.update { it.copy(isLoadingBalances = true) }
        viewModelScope.launch {
            runCatching { useCase.execute(current) }
                .onSuccess { balances = it }
                .onFailure { e -> _uiState.update { it.copy(error = e.message ?: "Failed to load balances") } }
            _uiState.update { it.copy(isLoadingBalances = false) }
            recompute()
        }
    }

    fun onQueryChange(query: String) {
        _uiState.update { it.copy(query = query) }
        recompute()
    }

    fun onFilterChange(filter: EntitiesFilter) {
        _uiState.update { it.copy(filter = filter) }
        recompute()
    }

    fun archive(entity: Entity) {
        val useCase = archiveUseCase ?: return
        val current = session ?: return
        viewModelScope.launch {
            runCatching { useCase.execute(current, entity) }
                .onFailure { e -> _uiState.update { it.copy(error = e.message ?: "Archive failed") } }
            // the snapshot listener refreshes the list
        }
    }

    /** Create (existingId == null) or update a party. Returns true on success. */
    fun save(input: EntityInput, existingId: String?, onDone: (Boolean) -> Unit) {
        val useCase = saveUseCase ?: return onDone(false)
        val current = session ?: return onDone(false)
        _uiState.update { it.copy(saving = true, saveError = null) }
        viewModelScope.launch {
            val result = runCatching { useCase.execute(current, input, existingId) }
            _uiState.update {
                it.copy(saving = false, saveError = result.exceptionOrNull()?.message)
            }
            refreshBalances()
            onDone(result.isSuccess)
        }
    }

    private fun recompute() {
        val query = _uiState.value.query.trim()
        val filter = _uiState.value.filter
        val rows = entities
            .asSequence()
            .filter { entity ->
                when (filter) {
                    EntitiesFilter.ALL -> true
                    EntitiesFilter.CUSTOMER -> EntityRole.CUSTOMER in entity.roles
                    EntitiesFilter.SUPPLIER -> EntityRole.SUPPLIER in entity.roles
                    EntitiesFilter.MIDDLEMAN -> EntityRole.MIDDLEMAN in entity.roles
                }
            }
            .filter { entity ->
                query.isBlank() ||
                    entity.name.contains(query, ignoreCase = true) ||
                    entity.phones.any { it.contains(query, ignoreCase = true) }
            }
            .map { EntityRow(it, balances[it.id]) }
            .toList()
        _uiState.update { it.copy(rows = rows) }
    }

    override fun onCleared() {
        // Carry-over from T1 review: release the shared Ktor HttpClient this repo owns.
        ledgerRepo?.close()
        super.onCleared()
    }
}
